scrollytelling/src/components/debugger/visualizer/index.tsx

```diff
@@ -0,0 +1,392 @@
+      // call a function whenever the cursor moves:
+      document.addEventListener("mousemove", elementDrag);
+    };
+
+    const elementDrag = (e: MouseEvent) => {
+      if (!panel) return;
+
+      e = e || window.event;
+      e.preventDefault();
+      // calculate the new cursor position:
+      pos1 = pos3 - e.clientX;
+      pos2 = pos4 - e.clientY;
+      pos3 = e.clientX;
+      pos4 = e.clientY;
+      // set the element's new position:
+      panel.style.top = panel.offsetTop - pos2 + "px";
+      panel.style.left = panel.offsetLeft - pos1 + "px";
+    };
+
+    function closeDragElement() {
+      /* stop moving when mouse button is released */
+      document.removeEventListener("mouseup", closeDragElement);
+      document.removeEventListener("mousemove", elementDrag);
+    }
+
+    panelHeader.addEventListener("mousedown", dragMouseDown);
+
+    return () => {
+      panelHeader.removeEventListener("mousedown", dragMouseDown);
+      closeDragElement();
+    };
+  }, []);
+
+  useEffect(() => {
+    return internalEventEmmiter.on("timeline:refresh", () => {
+      const roots: VisualizerRoot[] = [];
+
+      gsap.globalTimeline.getChildren().forEach((t) => {
+        if (!t.data?.isScrollytellingTween) return; // not a scrollytelling tween
+        const data = t.data as DataAttribute;
+        switch (data.type) {
+          case "root": {
+            const existingRootItem = roots.find((r) => r.id === data.id);
+            if (!existingRootItem) {
+              roots.push({
+                id: data.id,
+                debug: data.debug,
+                label: data.label,
+                children: [],
+                tween: t as any,
+              });
+            } else {
+              existingRootItem.debug = data.debug;
+              existingRootItem.tween = t as any;
+              existingRootItem.label = data.label;
+            }
+            break;
+          }
+          case "rest":
+          case "waypoint":
+          case "animation": {
+            const existingRootItem = roots.find((r) => r.id === data.rootId);
+            if (!existingRootItem) {
+              roots.push({
+                id: data.rootId,
+                debug: false,
+                label: data.rootId,
+                children: [t as any],
+                tween: undefined,
+              });
+            } else {
+              existingRootItem.children.push(t as any);
+            }
+
+            break;
+          }
+
+          default:
+            break;
+        }
+      });
+
+      // sort by trigger's top distance to top of the document
+      roots.sort((a, b) => {
+        const aTriggerEl = a.tween?.scrollTrigger?.trigger;
+        const bTriggerEl = b.tween?.scrollTrigger?.trigger;
+
+        if (!aTriggerEl || !bTriggerEl) return 0;
+
+        const aTop = aTriggerEl.getBoundingClientRect().top;
+        const bTop = bTriggerEl.getBoundingClientRect().top;
```

Comment:
@matiasperz thoughts on this? is the `getBoundingClientRect().top` a reliable way of getting the distance from the element to the top? i remember some nesting issues (and that's why react-use-measure was invented), but don't recall.